index.js

```diff
@@ -1,86 +1,64 @@
-          }))
-        }
+/**
+ * Get a kernel resources object 
+ * @param  {object}   kernelInfo              description of a kernel
+ * @param  {string}   kernelInfo.name         name of the kernel
+ * @param  {string}   kernelInfo.resourceDir  kernel's resources directory
+ * @return {Promise<object>}                  Promise for a kernelResources object
+ */
+function getKernelResources(kernelInfo) {
+  return Promise.resolve(kernelInfo).then(kernelInfo => 
+    callPromise(fs.readdir, [kernelInfo.resourceDir]).then(files => {
+      var kernelJSONIndex = files.indexOf('kernel.json');
+      if (kernelJSONIndex === -1) throw new Error('kernel.json not found');
+      
+      return callPromise(fs.readFile, [path.join(kernelInfo.resourceDir, 'kernel.json')]).then(data => ({
+        name: kernelInfo.name,
+        files: files.map(x => path.join(kernelInfo.resourceDir, x)),
+        resources_dir: kernelInfo.resourceDir,
+        spec: JSON.parse(data)
+      }));
     })
+  );
 }
 
-const kernelsBound = Observable.bindCallback(kernels)
-const getKernelResourcesBound = Observable.bindCallback(getKernelResources)
-
-function asObservable() {
-  return Observable.fromPromise(jp.dataDirs({ withSysPrefix: true }))
-                    .flatMap(x => {
-                      return Observable.fromArray(x)
-                    })
-                    .map(d => path.join(d, 'kernels'))
-                    .flatMap(x => {return kernelsBound(x)})
-                    .filter(x => x !== {})
-                    .flatMap(x => {
-                      return Observable.fromArray(x)
-                    })
-                    .flatMap(x => {
-                      return getKernelResourcesBound(x)
-                    })
-                    .filter(x => !x.err)
-                    .publish()
-                    .refCount()
+/**
+ * Gets a list of kernelInfo objects for a given directory of kernels
+ * @param  {string}   directory path to a directory full of kernels
+return return     files.map(fileName => ({
+      name: fileName,
+      resourceDir: path.join(directory, fileName)
+    }))
+  );
 }
 
-function asPromise() {
-  return asObservable()
-           .reduce((kernels, kernel) => {
-             kernels[kernel.name] = kernel;
-              return kernels;
-            }, {})
-           .toPromise()
+/**
+ * Get an array of kernelResources objects for the host environment
+ * @return {Promise<object[]>} Promise for an array of kernelResources objects
+ */
+function kernelSpecs() {
+  return jp.dataDirs({ withSysPrefix: true }).then(dirs => {
+    return Promise.all(dirs
+      .map(dir => getKernelInfos(dir)) // get kernel infos for each directory
+      .reduce((a, b) => a.concat(b)) // flatten the results into one array
+      .map(a => a.catch())) // ignore kernelInfo related errors
+      .map(a => getKernelResources(a)) // convert kernelInfo -> kernelResources
+      .map(a => a.catch()) // ignore kernelResources related errors
```

Comment:
This is why I wish for streams in general - stream of errors and a stream of valid data.